packages/ts/models/src/ConstraintBuilder.ts

```diff
@@ -0,0 +1,111 @@
+import type { EmptyObject } from 'type-fest';
+import {
+  $assertSupportedModel,
+  type Constraint,
+  type ConstraintFn,
+  type NonAttributedConstraint,
+  $defaultValue,
+  $name,
+  type AnyObject,
+  Model,
+  type Value,
+} from './Model.js';
+
+/**
+ * A builder class for declaring constraints.
+ * @internal low-level API for internal use only.
+ *
+ * @typeParam V - The value type of the constraint.
+ * @typeParam N - The name of the constraint.
+ * @typeParam A - The attributes for the constraint function. It is an object with string
+ * keys and arbitrary values. When the `value` attribute is the only one,
+ * the value can be given directly as the first argument of the constraint
+ * function; otherwise the attribute object should be provided.
+ */
+export class ConstraintBuilder<V = unknown, const N extends string = string, A extends AnyObject = EmptyObject> {
+  private supportedModel: Model<V>;
+  private [$name]: N | undefined;
+  private readonly attributeDefaults: Required<A>;
```

Comment:
```suggestion
  #supportedModel: Model<V>;
  protected [$name]: N | undefined;
  readonly #attributeDefaults: Required<A>;
```
How did you do that... I thought, `private`/`public` is forbidden by our linter...

Also, according to the style we have all over the project, symbol means `protected`, so it is the only one to use this modifier with.